test/archive.cpp

```diff
@@ -335,6 +335,69 @@ TEST(ZimArchive, cacheDontImpactReading)
   }
 }
 
+
+TEST(ZimArchive, cacheChange)
+{
+  for (auto& testfile: getDataFilePath("wikibooks_be_all_nopic_2017-02.zim")) {
+    auto archive = zim::Archive(testfile.path);
+
+    archive.set_dirent_cache_max_size(30);
+    archive.set_cluster_cache_max_size(5);
+
+    auto range = archive.iterEfficient();
+    auto it = range.begin();
+    for (auto i = 0; i<50 && it != range.end(); i++, it++) {
+      // Be sure to search by path to populate the dirent cache
+      auto entry = archive.getEntryByPath(it->getPath());
+      auto item = entry.getItem(true);
+      auto data = item.getData();
+    }
+    EXPECT_EQ(archive.get_dirent_cache_current_size(), 30);
+    EXPECT_EQ(archive.get_cluster_cache_current_size(), 2); // Only 2 clusters in the file
+
+    // Reduce cache size
+    archive.set_dirent_cache_max_size(10);
+    archive.set_cluster_cache_max_size(1);
+
+    EXPECT_EQ(archive.get_dirent_cache_current_size(), 10);
+    EXPECT_EQ(archive.get_cluster_cache_current_size(), 1);
+
+    for (auto i = 0; i<50 && it != range.end(); i++, it++) {
+      // Be sure to search by path to populate the dirent cache
+      auto entry = archive.getEntryByPath(it->getPath());
+      auto item = entry.getItem(true);
+      auto data = item.getData();
+    }
+
+    EXPECT_EQ(archive.get_dirent_cache_current_size(), 10);
+    EXPECT_EQ(archive.get_cluster_cache_current_size(), 1);
+
+    // Clean cache
+    // (More than testing the value, this is needed as we want to be sure the cache is actually populated later)
+    archive.set_dirent_cache_max_size(0);
+    archive.set_cluster_cache_max_size(0);
+
+    EXPECT_EQ(archive.get_dirent_cache_current_size(), 0);
+    EXPECT_EQ(archive.get_cluster_cache_current_size(), 0);
+
+    // Increase the cache
+    archive.set_dirent_cache_max_size(20);
+    archive.set_cluster_cache_max_size(1);
+    EXPECT_EQ(archive.get_dirent_cache_current_size(), 0);
+    EXPECT_EQ(archive.get_cluster_cache_current_size(), 0);
+
+    for (auto it = range.begin(); it != range.end(); it++) {
+      // Be sure to search by path to populate the dirent cache
+      auto entry = archive.getEntryByPath(it->getPath());
+      auto item = entry.getItem(true);
+      auto data = item.getData();
+    }
+
+    EXPECT_EQ(archive.get_dirent_cache_current_size(), 20);
+    EXPECT_EQ(archive.get_cluster_cache_current_size(), 1);
+  }
+}
```

Comment:
In this test, you only test for the absence of any effect of the changed cache size on operation indirectly - if the test doesn't crash, then the caching is supposed to work OK. However you may start getting wrong data because of having messed with caches!

Reply:
Equivalence of the archive (with different cache value) is already tested in the previous test. What is not tested is indeed that live changing the cache limit doesn't change the values.